# Initial data exploration for my datasets

# Note: this notebook is just for Noah's purposes and is not tidied up

More information about this project is available in my github repo here: https://github.com/Noah-Baustin/sf_crime_data_analysis

In [1]:
#import modules
import pandas as pd
#import altair as alt

/Users/nbaustin/.pyenv/versions/3.8.5/envs/sf_crime_data_analysis-3.8.5/lib/python3.8/site-packages/pandas/compat/__init__.py:124: UserWarning: Could not import the lzma module. Your installed Python is incomplete. Attempting to use lzma compression will result in a RuntimeError.
  warnings.warn(msg)


In [2]:
# import csv into a variable
historical_data = pd.read_csv('raw_data/SFPD_Incident_Reports_2003-May2018/Police_Department_Incident_Reports__Historical_2003_to_May_2018.csv', dtype=str)

Let's take a look at our data:

In [3]:
historical_data

PdId IncidntNum Incident Code        Category  \
0         3114751606302  031147516         06302   LARCENY/THEFT   
1         5069701104134  050697011         04134         ASSAULT   
2         6074729204104  060747292         04104         ASSAULT   
3         7103536315201  071035363         15201         ASSAULT   
4        11082415274000  110824152         74000  MISSING PERSON   
...                 ...        ...           ...             ...   
2129520  16099543005043  160995430         05043        BURGLARY   
2129521  16093783009031  160937830         09031  OTHER OFFENSES   
2129522  17078486916110  170784869         16110   DRUG/NARCOTIC   
2129523  18014823304138  180148233         04138         ASSAULT   
2129524  16044156851040  160441568         51040    NON-CRIMINAL   

                                                 Descript  DayOfWeek  \
0                             PETTY THEFT FROM A BUILDING     Sunday   
1                                                 BATTERY  Wednesday   
2                                                 ASSAULT   Saturday   
3                                                STALKING    Tuesday   
4                                           MISSING ADULT   Saturday   
...                                                   ...        ...   
2129520             BURGLARY OF RESIDENCE, UNLAWFUL ENTRY   Thursday   
2129521  MONEY, PROPERTY OR LABOR, FRAUDULENTLY OBTAINING     Monday   
2129522                    POSSESSION OF HEROIN FOR SALES    Tuesday   
2129523     BATTERY, FORMER SPOUSE OR DATING RELATIONSHIP   Saturday   
2129524                                        AIDED CASE     Monday   

               Date   Time PdDistrict      Resolution  ...  \
0        09/28/2003  10:00   SOUTHERN            NONE  ...   
1        06/22/2005  12:20   NORTHERN            NONE  ...   
2        07/15/2006  00:55    CENTRAL            NONE  ...   
3        09/25/2007  00:01    TARAVAL            NONE  ...   
4        09/24/2011  11:00    TARAVAL         LOCATED  ...   
...             ...    ...        ...             ...  ...   
2129520  12/01/2016  17:30   NORTHERN            NONE  ...   
2129521  11/14/2016  00:01  INGLESIDE            NONE  ...   
2129522  09/26/2017  01:38    CENTRAL  ARREST, BOOKED  ...   
2129523  02/24/2018  20:30    CENTRAL  ARREST, BOOKED  ...   
2129524  05/30/2016  11:33   SOUTHERN            NONE  ...   

        Fix It Zones as of 2017-11-06  2 2 DELETE - HSOC Zones 2 2  \
0                                      NaN                     NaN   
1                                      NaN                     NaN   
2                                      NaN                     NaN   
3                                      NaN                     NaN   
4                                      NaN                     NaN   
...                                    ...                     ...   
2129520                                NaN                       1   
2129521                                NaN                     NaN   
2129522                                 23                     NaN   
2129523                                NaN                     NaN   
2129524                                NaN                       1   

        Fix It Zones as of 2018-02-07 2 2  \
0                                     NaN   
1                                     NaN   
2                                     NaN   
3                                     NaN   
4                                     NaN   
...                                   ...   
2129520                               NaN   
2129521                               NaN   
2129522                                23   
2129523                               NaN   
2129524                               NaN   

        CBD, BID and GBD Boundaries as of 2017 2 2  \
0                                              NaN   
1                                              NaN   
2                                              N

In [4]:
historical_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2129525 entries, 0 to 2129524
Data columns (total 35 columns):
 #   Column                                                    Dtype 
---  ------                                                    ----- 
 0   PdId                                                      object
 1   IncidntNum                                                object
 2   Incident Code                                             object
 3   Category                                                  object
 4   Descript                                                  object
 5   DayOfWeek                                                 object
 6   Date                                                      object
 7   Time                                                      object
 8   PdDistrict                                                object
 9   Resolution                                                object
 10  Address                                   

I also see that I've got some date columns that need to be reformatted: for sure 'Date'. Let's reformat that:

In [5]:
historical_data['Date'] = pd.to_datetime(historical_data['Date'])

In [6]:
historical_data.head()

PdId IncidntNum Incident Code        Category  \
0   3114751606302  031147516         06302   LARCENY/THEFT   
1   5069701104134  050697011         04134         ASSAULT   
2   6074729204104  060747292         04104         ASSAULT   
3   7103536315201  071035363         15201         ASSAULT   
4  11082415274000  110824152         74000  MISSING PERSON   

                      Descript  DayOfWeek       Date   Time PdDistrict  \
0  PETTY THEFT FROM A BUILDING     Sunday 2003-09-28  10:00   SOUTHERN   
1                      BATTERY  Wednesday 2005-06-22  12:20   NORTHERN   
2                      ASSAULT   Saturday 2006-07-15  00:55    CENTRAL   
3                     STALKING    Tuesday 2007-09-25  00:01    TARAVAL   
4                MISSING ADULT   Saturday 2011-09-24  11:00    TARAVAL   

  Resolution  ... Fix It Zones as of 2017-11-06  2 2 DELETE - HSOC Zones 2 2  \
0       NONE  ...                                NaN                     NaN   
1       NONE  ...                                NaN                     NaN   
2       NONE  ...                                NaN                     NaN   
3       NONE  ...                                NaN                     NaN   
4    LOCATED  ...                                NaN                     NaN   

  Fix It Zones as of 2018-02-07 2 2  \
0                               NaN   
1                               NaN   
2                               NaN   
3                               NaN   
4                               NaN   

  CBD, BID and GBD Boundaries as of 2017 2 2 Areas of Vulnerability, 2016 2 2  \
0                                        NaN                              NaN   
1                                        NaN                                2   
2                                        NaN                                2   
3                                        NaN                                1   
4                                        NaN                              NaN   

  Central Market/Tenderloin Boundary 2 2  \
0                                    NaN   
1                                    NaN   
2                                    NaN   
3                                    NaN   
4                                    NaN   

  Central Market/Tenderloin Boundary Polygon - Updated 2 2  \
0                                                NaN         
1                                                NaN         
2                                                NaN         
3                                                NaN         
4                                                NaN         

  HSOC Zones as of 2018-06-05 2 2 OWED Public Spaces 2 2 Neighborhoods 2  
0                             NaN                    NaN             NaN  
1                             NaN                    NaN              97  
2                             NaN                    NaN             106  
3                             NaN                    NaN              49  
4                             NaN                    NaN             NaN  

[5 rows x 35 columns]

Let's take a closer look at our columns:

In [7]:
historical_data.columns

Index(['PdId', 'IncidntNum', 'Incident Code', 'Category', 'Descript',
       'DayOfWeek', 'Date', 'Time', 'PdDistrict', 'Resolution', 'Address', 'X',
       'Y', 'location', 'SF Find Neighborhoods 2 2',
       'Current Police Districts 2 2', 'Current Supervisor Districts 2 2',
       'Analysis Neighborhoods 2 2', 'DELETE - Fire Prevention Districts 2 2',
       'DELETE - Police Districts 2 2', 'DELETE - Supervisor Districts 2 2',
       'DELETE - Zip Codes 2 2', 'DELETE - Neighborhoods 2 2',
       'DELETE - 2017 Fix It Zones 2 2',
       'Civic Center Harm Reduction Project Boundary 2 2',
       'Fix It Zones as of 2017-11-06  2 2', 'DELETE - HSOC Zones 2 2',
       'Fix It Zones as of 2018-02-07 2 2',
       'CBD, BID and GBD Boundaries as of 2017 2 2',
       'Areas of Vulnerability, 2016 2 2',
       'Central Market/Tenderloin Boundary 2 2',
       'Central Market/Tenderloin Boundary Polygon - Updated 2 2',
       'HSOC Zones as of 2018-06-05 2 2', 'OWED Public Spaces 2 2',
       

We can see here that there's a bunch of extra columns in the data that's not included in the documentation and are not essential to our analysis. Let's get rid of those columns.

In [8]:
#get rid of all those extra columns we don't need

historical_data = historical_data[['PdId', 'IncidntNum', 'Incident Code', 'Category', 'Descript',
       'DayOfWeek', 'Date', 'Time', 'PdDistrict', 'Resolution', 'Address', 'X',
       'Y', 'location']].copy()

In [9]:
historical_data.columns

Index(['PdId', 'IncidntNum', 'Incident Code', 'Category', 'Descript',
       'DayOfWeek', 'Date', 'Time', 'PdDistrict', 'Resolution', 'Address', 'X',
       'Y', 'location'],
      dtype='object')

Now that our columns are cleaned up and our dates are formatted correctly, let's take a look at our date range included in the data:

In [10]:
historical_data['Date'].min()

Timestamp('2003-01-01 00:00:00')

In [11]:
historical_data['Date'].max()

Timestamp('2018-05-15 00:00:00')

We can see here that it appears that we most likely have complete data beginning in 2003. BUT we see here that the 2018 data is incomplete, so if we want to do an annual analysis, we'll need to exclude 2018.

We've got 2,129,525 entries. Let's check out our three different columns that have incident indentification codes to see how many unique values are in each:

In [12]:
historical_data['IncidntNum'].nunique()

1703626

In [13]:
historical_data['Incident Code'].nunique()

885

In [14]:
historical_data['PdId'].nunique()

2129525

Our documentation tells us that the PdId column is equivelant to the row_id column in the new data (see downloaded pdf titled: Change Notice - Police Incident Reports). And the documentation for the newer dataset tells us that the row_id is the unique identifier for each row. 

It's a good sign that there's exaclty as many PdId unique values as there are rows in the dataset.

It also makes sense that there are duplicate values in the IncidntNum column. The IncidntNum refers to the case number. So if a supplemental report was filed after the incident was initially entered in this dataset, it would show up with a new PdId BUT the IncidntNum would be the same. 

But that does mean that I need to make sure I'm not counting the same incident multiple times if it shows up in this dataset multiple times. Most likely, my analysis will focus on a unique set of IncidntNum. 

Let's double check that there's no duplicate rows:

In [15]:
historical_data[historical_data.duplicated()]

Empty DataFrame
Columns: [PdId, IncidntNum, Incident Code, Category, Descript, DayOfWeek, Date, Time, PdDistrict, Resolution, Address, X, Y, location]
Index: []

No duplicate rows, that's great.

Let's find out how many duplicated IncidntNum values we have:

In [16]:
historical_data[historical_data['IncidntNum'].duplicated()]

PdId IncidntNum Incident Code        Category  \
523      17001676971010  170016769         71010  OTHER OFFENSES   
562      18023965028150  180239650         28150       VANDALISM   
857      17018698663010  170186986         63010        WARRANTS   
968      17063457109320  170634571         09320           FRAUD   
1179     16096494626135  160964946         26135         ASSAULT   
...                 ...        ...           ...             ...   
2129517  16092245574000  160922455         74000  MISSING PERSON   
2129519  18014350616652  180143506         16652   DRUG/NARCOTIC   
2129521  16093783009031  160937830         09031  OTHER OFFENSES   
2129522  17078486916110  170784869         16110   DRUG/NARCOTIC   
2129524  16044156851040  160441568         51040    NON-CRIMINAL   

                                                 Descript DayOfWeek  \
523                             LOST/STOLEN LICENSE PLATE  Thursday   
562                         MALICIOUS MISCHIEF, VANDALISM  Saturday   
857                                        WARRANT ARREST    Monday   
968                          CREDIT CARD, THEFT BY USE OF    Friday   
1179         UNLAWFUL DISSUADING/THREATENING OF A WITNESS    Sunday   
...                                                   ...       ...   
2129517                                     MISSING ADULT  Saturday   
2129519           POSSESSION OF METH-AMPHETAMINE FOR SALE  Thursday   
2129521  MONEY, PROPERTY OR LABOR, FRAUDULENTLY OBTAINING    Monday   
2129522                    POSSESSION OF HEROIN FOR SALES   Tuesday   
2129524                                        AIDED CASE    Monday   

              Date   Time  PdDistrict      Resolution  \
523     2017-01-05  18:00     BAYVIEW            NONE   
562     2018-03-31  14:30     TARAVAL            NONE   
857     2017-03-06  16:00  TENDERLOIN  ARREST, BOOKED   
968     2017-08-04  16:00     CENTRAL            NONE   
1179    2016-11-27  14:29     BAYVIEW  ARREST, BOOKED   
...            ...    ...         ...             ...   
2129517 2016-11-12  15:08    SOUTHERN            NONE   
2129519 2018-02-22  23:27     CENTRAL  ARREST, BOOKED   
2129521 2016-11-14  00:01   INGLESIDE            NONE   
2129522 2017-09-26  01:38     CENTRAL  ARREST, BOOKED   
2129524 2016-05-30  11:33    SOUTHERN            NONE   

                             Address                    X                   Y  \
523          1300 Block of THOMAS AV  -122.38667457883558   37.72772896388117   
562         0 Block of THOMASMORE WY  -122.47293744666878  37.712252275226234   
857      200 Block of LEAVENWORTH ST  -122.41408113801357  37.783276726606246   
968            KEARNY ST / MARKET ST  -122.40343408674543   37.78764305017998   
1179        0 Block of NORTHRIDGE RD  -122.37560517196246   37.73072152157033   
...                              ...                  ...                 ...   
2129517       800 Block of BRYANT ST  -122.40340479147905     37.775420706711   
2129519         0 Block of MAIDEN LN  -122.40435958710997  37.788272654251344   
2129521        0 Block of CONCORD ST  -122.44276988992716   37.71275870748528   
2129522     500 Block of BROADWAY ST  -122.40586321263237   37.79802284521844   
2129524      1600 Block of MARKET ST   -122.4209584484125  37.773955543024975   

                                               location  
523       POINT (-122.38667457883558 37.72772896388117)  
562      POINT (-122.47293744666878 37.712252275226234)  
857      POINT (-122.41408113801357 37.783276726606246)  
968       POINT (-122.40343408674543 37.78764305017998)  
1179      POINT (-122.37560517196246 37.73072152157033)  
...                                                 ...  
2129517     POINT (-122.40340479147905 37.775420706711)  
2129519  POINT (-122.40435958710997 37.788272654251344)  
2129521   POINT (-122.44276988992716 37.71275870748528)  
2129522   POINT (-122.40586321263237 37.79802284521844)  
2129524   POINT (-122.4209584484125 37.77395554302497

In [17]:
historical_data['IncidntNum'].nunique() + len(historical_data[historical_data['IncidntNum'].duplicated()])

2129525

There's 425,899 duplicates for 'IncidntNum'. We we add that to the number of unique values (above) it is equal to the number of rows in our dataset... that tells us that 1,703,626 is the actual number of incidents that we're working with.

#### NOTE: explain this step better b/c confusion in Soo meeting

Let's take a look at some of our duplicates:

First we'll create a dataframe with our duplicated cases:

In [18]:
dupe_cases = historical_data[historical_data['IncidntNum'].duplicated()].copy()

Now we create a `list` or `array` (the numpy version of a list) of those unique PdId's:

In [19]:
dupe_IncidentNum = dupe_cases['IncidntNum'].to_list()

Finally, we're displaying a subset our of original data that just includes the duplicate incident numbers:

In [20]:
dupe_cases_full = historical_data[historical_data['IncidntNum'].isin(dupe_IncidentNum)].sort_values(by='IncidntNum')

In [21]:
dupe_cases_full.head(30)

PdId IncidntNum Incident Code                Category  \
1021616    6238907023  000062389         07023           VEHICLE THEFT   
1869704    6238907043  000062389         07043           VEHICLE THEFT   
1003273   12343616010  000123436         16010           DRUG/NARCOTIC   
625462    12343602004  000123436         02004  SEX OFFENSES, FORCIBLE   
1836290   24825063010  000248250         63010                WARRANTS   
1009328   24825062050  000248250         62050                WARRANTS   
1554935   32468419090  000324684         19090             DRUNKENNESS   
1297782   32468463010  000324684         63010                WARRANTS   
1434901   36322862050  000363228         62050                WARRANTS   
1029718   36322863010  000363228         63010                WARRANTS   
797056    76724426080  000767244         26080          OTHER OFFENSES   
1734920   76724416621  000767244         16621           DRUG/NARCOTIC   
1238053   76724416700  000767244         16700           DRUG/NARCOTIC   
1279759   81020726142  000810207         26142              KIDNAPPING   
444741    81020704144  000810207         04144  SEX OFFENSES, FORCIBLE   
1362027   81020702204  000810207         02204  SEX OFFENSES, FORCIBLE   
490689    88865207023  000888652         07023           VEHICLE THEFT   
464553    88865207043  000888652         07043           VEHICLE THEFT   
802280    96275107023  000962751         07023           VEHICLE THEFT   
870203    96275107043  000962751         07043           VEHICLE THEFT   
1373007  100017663010  001000176         63010                WARRANTS   
1227820  100017662050  001000176         62050                WARRANTS   
484782   124909527065  001249095         27065                   FRAUD   
886947   124909563010  001249095         63010                WARRANTS   
1793335  132654662020  001326546         62020                WARRANTS   
1563382  132654663010  001326546         63010                WARRANTS   
881525   144089562060  001440895         62060                WARRANTS   
1711046  144089563010  001440895         63010                WARRANTS   
1892575  313990175000  003139901         75000          MISSING PERSON   
893672   313990174000  003139901         74000          MISSING PERSON   

                                             Descript  DayOfWeek       Date  \
1021616                             STOLEN MOTORCYCLE    Tuesday 2004-03-30   
1869704                VEHICLE, RECOVERED, MOTORCYCLE    Tuesday 2004-03-30   
1003273                       POSSESSION OF MARIJUANA     Friday 2004-07-02   
625462                    FORCIBLE RAPE, BODILY FORCE     Friday 2004-07-02   
1836290                                WARRANT ARREST     Monday 2003-01-27   
1009328               ENROUTE TO OUTSIDE JURISDICTION     Monday 2003-01-27   
1554935  UNDER INFLUENCE OF ALCOHOL IN A PUBLIC PLACE     Monday 2003-03-31   
1297782                                WARRANT ARREST     Monday 2003-03-31   
1434901               ENROUTE TO OUTSIDE JURISDICTION  Wednesday 2004-11-03   
1029718                                WARRANT ARREST  Wednesday 2004-11-03   
797056                                     CONSPIRACY     Sunday 2003-06-29   
1734920               POSSESSION OF BASE/ROCK COCAINE     Sunday 2003-06-29   
1238053       VISITING WHERE DRUGS ARE USED OR SMOKED     Sunday 2003-06-29   
1279759                            FALSE IMPRISONMENT  Wednesday 2003-07-09   
444741                                 SEXUAL BATTERY  Wednesday 2003-07-09   
1362027                  ATTEMPTED RAPE, BODILY FORCE  Wednesday 2003-07-09   
490689                              STOLEN MOTORCYCLE    Tuesday 2004-06-15   
464553                 VEHICLE, RECOVERED, MOTORCYCLE    Tuesday 2004-06-15   
802280                              STOLEN MOTORCYCLE  Wednesday 2004-06-23   
870203                 VEHICLE, RECOVERED, MOTORCYCLE  Wednesday 2004-06-23   
1373007                                WARRANT ARREST     Monda

We can see here anecdotely that additional entries for duplicated IncidentNum's include additional entries in the descript column in some cases, but sometimes they remain the same. 
#### come back to this!

We're going to want to isolate our incidents that include marijuana crimes. So let's take a look at the unique values in the two columns that might contain information about marijuana crimes:

In [22]:
historical_data['Category'].unique()

array(['LARCENY/THEFT', 'ASSAULT', 'MISSING PERSON', 'VEHICLE THEFT',
       'BURGLARY', 'DRUG/NARCOTIC', 'DRIVING UNDER THE INFLUENCE',
       'VANDALISM', 'OTHER OFFENSES', 'DRUNKENNESS', 'NON-CRIMINAL',
       'ROBBERY', 'SUSPICIOUS OCC', 'TRESPASS', 'WARRANTS',
       'FORGERY/COUNTERFEITING', 'STOLEN PROPERTY',
       'SEX OFFENSES, FORCIBLE', 'FRAUD', 'SECONDARY CODES',
       'PROSTITUTION', 'RECOVERED VEHICLE', 'BRIBERY', 'ARSON',
       'DISORDERLY CONDUCT', 'WEAPON LAWS', 'LIQUOR LAWS', 'EXTORTION',
       'SUICIDE', 'KIDNAPPING', 'SEX OFFENSES, NON FORCIBLE',
       'BAD CHECKS', 'EMBEZZLEMENT', 'LOITERING', 'GAMBLING', 'TREA',
       'PORNOGRAPHY/OBSCENE MAT'], dtype=object)

We can see here that there's a DRUG/NARCOTIC category, so that's probably where we're going to find the marijuana crimes. But nothing specific about marijuana here. That's going to show up in our 'Descript' column. There's too many unique values in that column to list, so let's create a subset:

In [23]:
#create dataframe with all our marijuana incidents

historical_data_marijuana = historical_data[
    historical_data['Descript'].str.contains('MARIJUANA')
].copy()

## Soo meeting notes

In [24]:
drug_narcotic_incidents = historical_data[
    historical_data['Category'] == 'DRUG/NARCOTIC'
].reset_index(drop=True)

In [25]:
# needed if I decide to answer my 'extra' question: Compare marijuana arrests to other types of crimes, like narcotics.

#drug_narcotic_incidents.to_csv('drug_narcotic_incidents_historical.csv', index=False)

In [26]:
drug_narcotic_incidents.head()

PdId IncidntNum Incident Code       Category  \
0  18032260316100  180322603         16100  DRUG/NARCOTIC   
1  16035574316650  160355743         16650  DRUG/NARCOTIC   
2  17001952516623  170019525         16623  DRUG/NARCOTIC   
3  17089965616662  170899656         16662  DRUG/NARCOTIC   
4  16055139916010  160551399         16010  DRUG/NARCOTIC   

                                      Descript DayOfWeek       Date   Time  \
0                         POSSESSION OF HEROIN   Tuesday 2018-05-01  13:27   
1               POSSESSION OF METH-AMPHETAMINE    Sunday 2016-05-01  08:55   
2     POSSESSION OF BASE/ROCK COCAINE FOR SALE  Saturday 2017-01-07  23:50   
3  POSSESSION OF CONTROLLED SUBSTANCE FOR SALE    Friday 2017-11-03  21:00   
4                      POSSESSION OF MARIJUANA    Friday 2016-07-08  08:00   

   PdDistrict      Resolution                    Address                    X  \
0    SOUTHERN  ARREST, BOOKED  600 Block of STEVENSON ST  -122.41265189435406   
1  TENDERLOIN  ARREST, BOOKED   LEAVENWORTH ST / TURK ST  -122.41405629189116   
2    NORTHERN  ARREST, BOOKED   2100 Block of LOMBARD ST  -122.43686333760363   
3  TENDERLOIN  ARREST, BOOKED        0 Block of JONES ST  -122.41219255572126   
4     MISSION  ARREST, BOOKED       0 Block of ALBION ST  -122.42326589360349   

                   Y                                       location  
0  37.77951316688882  POINT (-122.41265189435406 37.77951316688882)  
1  37.78279310710062  POINT (-122.41405629189116 37.78279310710062)  
2  37.79961996899933  POINT (-122.43686333760363 37.79961996899933)  
3  37.78160457412525  POINT (-122.41219255572126 37.78160457412525)  
4    37.765649515945    POINT (-122.42326589360349 37.765649515945)

In [27]:
#how to isolate the marijuana duplicate incidentnum's
historical_data[
    historical_data['Descript'].str.contains('MARIJUANA') & historical_data['IncidntNum'].isin(dupe_IncidentNum)
].sort_values(by='IncidntNum')

PdId IncidntNum Incident Code       Category  \
1003273     12343616010  000123436         16010  DRUG/NARCOTIC   
1222036   1107270116010  011072701         16010  DRUG/NARCOTIC   
1576579   3000127216010  030001272         16010  DRUG/NARCOTIC   
1412714   3000319316030  030003193         16030  DRUG/NARCOTIC   
1289479   3000319316040  030003193         16040  DRUG/NARCOTIC   
...                 ...        ...           ...            ...   
2044767  18031791116050  180317911         16050  DRUG/NARCOTIC   
657867   18032993816010  180329938         16010  DRUG/NARCOTIC   
2081965  18034889716060  180348897         16060  DRUG/NARCOTIC   
672658   18034889716030  180348897         16030  DRUG/NARCOTIC   
1314490  96052487916030  960524879         16030  DRUG/NARCOTIC   

                                  Descript  DayOfWeek       Date   Time  \
1003273            POSSESSION OF MARIJUANA     Friday 2004-07-02  14:08   
1222036            POSSESSION OF MARIJUANA   Saturday 2011-09-10  01:00   
1576579            POSSESSION OF MARIJUANA  Wednesday 2003-01-01  06:32   
1412714  POSSESSION OF MARIJUANA FOR SALES  Wednesday 2003-01-01  19:52   
1289479                  SALE OF MARIJUANA  Wednesday 2003-01-01  19:52   
...                                    ...        ...        ...    ...   
2044767               FURNISHING MARIJUANA     Sunday 2018-04-29  17:36   
657867             POSSESSION OF MARIJUANA   Thursday 2018-05-03  23:14   
2081965        TRANSPORTATION OF MARIJUANA   Thursday 2018-05-10  15:30   
672658   POSSESSION OF MARIJUANA FOR SALES   Thursday 2018-05-10  15:30   
1314490  POSSESSION OF MARIJUANA FOR SALES    Tuesday 2012-11-27  09:00   

         PdDistrict      Resolution                          Address  \
1003273    SOUTHERN  ARREST, BOOKED           800 Block of BRYANT ST   
1222036  TENDERLOIN  ARREST, BOOKED  LEAVENWORTH ST / GOLDEN GATE AV   
1576579    SOUTHERN  ARREST, BOOKED              400 Block of 6TH ST   
1412714     MISSION  ARREST, BOOKED         2100 Block of MISSION ST   
1289479     MISSION  ARREST, BOOKED         2100 Block of MISSION ST   
...             ...             ...                              ...   
2044767        PARK  ARREST, BOOKED         HAIGHT ST / BELVEDERE ST   
657867   TENDERLOIN  ARREST, BOOKED         600 Block of OFARRELL ST   
2081965    SOUTHERN  ARREST, BOOKED           600 Block of MARKET ST   
672658     SOUTHERN  ARREST, BOOKED           600 Block of MARKET ST   
1314490  TENDERLOIN            NONE             0 Block of POWELL ST   

                           X                  Y  \
1003273    -122.403404791479    37.775420706711   
1222036    -122.413869632554   37.7818621883021   
1576579    -122.402983634198   37.7763280295782   
1412714    -122.419470488515   37.7622701739914   
1289479    -122.419470488515   37.7622701739914   
...                      ...                ...   
2044767  -122.44964766295456  37.76966191958464   
657867   -122.41597762782685  37.78546705050173   
2081965  -122.40255295786527  37.78847491468028   
672658   -122.40255295786527  37.78847491468028   
1314490    -122.407845374111   37.7850943210103   

                                              location  
1003273    POINT (-122.40340479147902 37.775420706711)  
1222036   POINT (-122.41386963255401 37.7818621883021)  
1576579   POINT (-122.40298363419801 37.7763280295782)  
1412714   POINT (-122.41947048851502 37.7622701739914)  
1289479   POINT (-122.41947048851502 37.7622701739914)  
...                                                ...  
2044767  POINT (-122.44964766295456 37.76966191958464)  
657867   POINT (-122.41597762782685 37.78546705050173)  
2081965  POINT (-122.40255295786527 37.78847491468028)  
672658   POINT (-122.40255295786527 37.78847491468028)  
1314490     POINT (-122.407845374111 37.7850943210103)  

[16010 rows x 14 columns]

In [28]:
historical_data[historical_data['IncidntNum'] == "000123436"]

PdId IncidntNum Incident Code                Category  \
625462   12343602004  000123436         02004  SEX OFFENSES, FORCIBLE   
1003273  12343616010  000123436         16010           DRUG/NARCOTIC   

                            Descript DayOfWeek       Date   Time PdDistrict  \
625462   FORCIBLE RAPE, BODILY FORCE    Friday 2004-07-02  14:08   SOUTHERN   
1003273      POSSESSION OF MARIJUANA    Friday 2004-07-02  14:08   SOUTHERN   

             Resolution                 Address                  X  \
625462   ARREST, BOOKED  800 Block of BRYANT ST  -122.403404791479   
1003273  ARREST, BOOKED  800 Block of BRYANT ST  -122.403404791479   

                       Y                                     location  
625462   37.775420706711  POINT (-122.40340479147902 37.775420706711)  
1003273  37.775420706711  POINT (-122.40340479147902 37.775420706711)

In [29]:
#show info about the first row in the dataframe [iloc means index location]
incident_duplicate.iloc[0]

NameError: name 'incident_duplicate' is not defined

In [ ]:
#what unique values are in this marijuana data frame for types of arrests
historical_data_marijuana['Descript'].unique()

In [ ]:
#dropping duplicate incident numbers
#going to need to write an explainer 
historical_data_marijuana[historical_data_marijuana['IncidntNum'].duplicated()]

In [ ]:
historical_data_marijuana[historical_data_marijuana['IncidntNum'] == '160676737']

In [ ]:
# export the df to a csv
#historical_data_marijuana.to_csv("historical_data_marijuana.csv", index=False)

### Bring in the more recent dataset

In [ ]:
newer_data.columns

In [ ]:
historical_data.columns

In [3]:
# import csv into a variable
newer_data = pd.read_csv('raw_data/SFPD_Incident_Reports_2018-10.14.21/Police_Department_Incident_Reports__2018_to_Present(1).csv', dtype=str)

In [ ]:
historical_data['Resolution'].unique()

In [4]:
newer_data.columns

Index(['Incident Datetime', 'Incident Date', 'Incident Time', 'Incident Year',
       'Incident Day of Week', 'Report Datetime', 'Row ID', 'Incident ID',
       'Incident Number', 'CAD Number', 'Report Type Code',
       'Report Type Description', 'Filed Online', 'Incident Code',
       'Incident Category', 'Incident Subcategory', 'Incident Description',
       'Resolution', 'Intersection', 'CNN', 'Police District',
       'Analysis Neighborhood', 'Supervisor District', 'Latitude', 'Longitude',
       'Point', 'Neighborhoods', 'ESNCAG - Boundary File',
       'Central Market/Tenderloin Boundary Polygon - Updated',
       'Civic Center Harm Reduction Project Boundary',
       'HSOC Zones as of 2018-06-05', 'Invest In Neighborhoods (IIN) Areas',
       'Current Supervisor Districts', 'Current Police Districts'],
      dtype='object')

In [54]:
newer_data['Incident Description'] = newer_data['Incident Description'].str.upper()

Now we need to figure out where the marijuana cases are organized in the newer dataset

In [55]:
find_marijuana_1 = newer_data[
    newer_data['Incident Description'].str.contains('MARIJUANA')
].reset_index(drop=True)

In [56]:
find_marijuana_1

Incident Datetime Incident Date Incident Time Incident Year  \
0    2021/04/21 09:22:00 AM    2021/04/21         09:22          2021   
1    2021/10/08 12:42:00 PM    2021/10/08         12:42          2021   
2    2021/10/09 12:26:00 PM    2021/10/09         12:26          2021   
3    2021/06/05 11:07:00 AM    2021/06/05         11:07          2021   
4    2021/07/07 10:01:00 PM    2021/07/07         22:01          2021   
..                      ...           ...           ...           ...   
532  2020/09/02 04:28:00 PM    2020/09/02         16:28          2020   
533  2020/10/22 08:47:00 AM    2020/10/22         08:47          2020   
534  2021/01/19 03:13:00 AM    2021/01/19         03:13          2021   
535  2021/01/17 10:05:00 AM    2021/01/17         10:05          2021   
536  2020/05/21 12:08:00 AM    2020/05/21         00:08          2020   

    Incident Day of Week         Report Datetime        Row ID Incident ID  \
0              Wednesday  2021/04/21 09:55:00 AM  103250716010     1032507   
1                 Friday  2021/10/08 12:43:00 PM  107857316010     1078573   
2               Saturday  2021/10/09 12:26:00 PM  107886216030     1078862   
3               Saturday  2021/06/05 11:07:00 AM  103724316030     1037243   
4              Wednesday  2021/07/07 10:01:00 PM  104811216030     1048112   
..                   ...                     ...           ...         ...   
532            Wednesday  2020/09/02 04:28:00 PM   95820816010      958208   
533             Thursday  2020/10/22 02:42:00 PM   97176416030      971764   
534              Tuesday  2021/01/19 03:13:00 AM   99731116030      997311   
535               Sunday  2021/01/17 10:05:00 AM   99688016030      996880   
536             Thursday  2020/05/21 12:08:00 AM   92864616030      928646   

    Incident Number CAD Number  ...            Longitude  \
0         210259178  210259178  ...  -122.40851633190513   
1         210656643  212811587  ...  -122.40667700592424   
2         210658815  212821487  ...  -122.38717421937969   
3         210348549  211561131  ...  -122.42353498869593   
4         210428470  211883301  ...  -122.44331168934727   
..              ...        ...  ...                  ...   
532       200528905  202462307  ...  -122.43129384491388   
533       200638140  202961857  ...   -122.5040487777567   
534       210041141  210190283  ...  -122.41811812009416   
535       210037532  210170877  ...  -122.42746205880601   
536       200308092  201420020  ...  -122.41561643757444   

                                              Point Neighborhoods  \
0     POINT (-122.40851633190513 37.77375969975922)            32   
1    POINT (-122.40667700592424 37.767142180962104)            33   
2     POINT (-122.38717421937969 37.74615712680034)            56   
3     POINT (-122.42353498869593 37.77223581387494)            26   
4     POINT (-122.44331168934727 37.77046953691186)           112   
..                                              ...           ...   
532   POINT (-122.43129384491388 37.72872954740928)            94   
533    POINT (-122.5040487777567 37.76426117163131)            39   
534    POINT (-122.41811812009416 37.7603010605011)            53   
535   POINT (-122.42746205880601 37.76877049785351)            28   
536   POINT (-122.41561643757444 37.80727586975037)            99   

    ESNCAG - Boundary File  \
0                      NaN   
1                      NaN   
2                      NaN   
3                      NaN   
4                      NaN   
..                     ...   
532                    NaN   
533                    NaN   
534                    NaN   
535                    NaN   
536                    NaN   

    Central Market/Tenderloin Boundary Polygon - Updated  \
0                                                  NaN     
1                                                  NaN     
2                                                  NaN     
3                                            

In [58]:
find_marijuana_1['Incident Description'].unique()

array(['MARIJUANA OFFENSE', 'MARIJUANA, POSSESSION FOR SALE',
       'MARIJUANA, SALES', 'MARIJUANA, CULTIVATING/PLANTING',
       'MARIJUANA, FURNISHING', 'MARIJUANA, TRANSPORTING'], dtype=object)

Unfortunately this shows us that the Incident Description column does not contain information categorizing crimes by the term 'marijuana', unlike the older data.

In [18]:
find_marijuana_2 = newer_data[
    newer_data['Incident Subcategory'].str.contains('MARIJUANA', na=False)
].reset_index(drop=True)

In [19]:
find_marijuana_2

Empty DataFrame
Columns: [Incident Datetime, Incident Date, Incident Time, Incident Year, Incident Day of Week, Report Datetime, Row ID, Incident ID, Incident Number, CAD Number, Report Type Code, Report Type Description, Filed Online, Incident Code, Incident Category, Incident Subcategory, Incident Description, Resolution, Intersection, CNN, Police District, Analysis Neighborhood, Supervisor District, Latitude, Longitude, Point, Neighborhoods, ESNCAG - Boundary File, Central Market/Tenderloin Boundary Polygon - Updated, Civic Center Harm Reduction Project Boundary, HSOC Zones as of 2018-06-05, Invest In Neighborhoods (IIN) Areas, Current Supervisor Districts, Current Police Districts]
Index: []

[0 rows x 34 columns]

No marijuana strings in the incident subcategory either.

In [20]:
find_marijuana_3 = newer_data[
    newer_data['Incident Category'].str.contains('MARIJUANA', na=False)
].reset_index(drop=True)

In [21]:
find_marijuana_3

Empty DataFrame
Columns: [Incident Datetime, Incident Date, Incident Time, Incident Year, Incident Day of Week, Report Datetime, Row ID, Incident ID, Incident Number, CAD Number, Report Type Code, Report Type Description, Filed Online, Incident Code, Incident Category, Incident Subcategory, Incident Description, Resolution, Intersection, CNN, Police District, Analysis Neighborhood, Supervisor District, Latitude, Longitude, Point, Neighborhoods, ESNCAG - Boundary File, Central Market/Tenderloin Boundary Polygon - Updated, Civic Center Harm Reduction Project Boundary, HSOC Zones as of 2018-06-05, Invest In Neighborhoods (IIN) Areas, Current Supervisor Districts, Current Police Districts]
Index: []

[0 rows x 34 columns]

In [22]:
find_marijuana_4 = newer_data[
    newer_data['Report Type Description'].str.contains('MARIJUANA', na=False)
].reset_index(drop=True)

In [23]:
find_marijuana_4

Empty DataFrame
Columns: [Incident Datetime, Incident Date, Incident Time, Incident Year, Incident Day of Week, Report Datetime, Row ID, Incident ID, Incident Number, CAD Number, Report Type Code, Report Type Description, Filed Online, Incident Code, Incident Category, Incident Subcategory, Incident Description, Resolution, Intersection, CNN, Police District, Analysis Neighborhood, Supervisor District, Latitude, Longitude, Point, Neighborhoods, ESNCAG - Boundary File, Central Market/Tenderloin Boundary Polygon - Updated, Civic Center Harm Reduction Project Boundary, HSOC Zones as of 2018-06-05, Invest In Neighborhoods (IIN) Areas, Current Supervisor Districts, Current Police Districts]
Index: []

[0 rows x 34 columns]

Frustratingly I'm not finding any marijuana information in any of these columns

In [24]:
upper_test = newer_data

In [25]:
upper_test = upper_test.upper()

AttributeError: 'DataFrame' object has no attribute 'upper'

In [26]:
test = 'hello'
print(test)

hello


In [27]:
test.upper()

'HELLO'

In [32]:
newer_data.columns

Index(['Incident Datetime', 'Incident Date', 'Incident Time', 'Incident Year',
       'Incident Day of Week', 'Report Datetime', 'Row ID', 'Incident ID',
       'Incident Number', 'CAD Number', 'Report Type Code',
       'Report Type Description', 'Filed Online', 'Incident Code',
       'Incident Category', 'Incident Subcategory', 'Incident Description',
       'Resolution', 'Intersection', 'CNN', 'Police District',
       'Analysis Neighborhood', 'Supervisor District', 'Latitude', 'Longitude',
       'Point', 'Neighborhoods', 'ESNCAG - Boundary File',
       'Central Market/Tenderloin Boundary Polygon - Updated',
       'Civic Center Harm Reduction Project Boundary',
       'HSOC Zones as of 2018-06-05', 'Invest In Neighborhoods (IIN) Areas',
       'Current Supervisor Districts', 'Current Police Districts'],
      dtype='object')

In [35]:
incident_descript_upper = newer_data['Incident Subcategory'].str.upper()

In [38]:
find_marijuana_5 = incident_descript_upper.str.contains('MARIJUANA').reset_index(drop=True)

In [41]:
find_marijuana_5.unique()

array([False, nan], dtype=object)

AttributeError: 'Series' object has no attribute 'info'

In [47]:
newer_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 513217 entries, 0 to 513216
Data columns (total 34 columns):
 #   Column                                                Non-Null Count   Dtype 
---  ------                                                --------------   ----- 
 0   Incident Datetime                                     513217 non-null  object
 1   Incident Date                                         513217 non-null  object
 2   Incident Time                                         513217 non-null  object
 3   Incident Year                                         513217 non-null  object
 4   Incident Day of Week                                  513217 non-null  object
 5   Report Datetime                                       513217 non-null  object
 6   Row ID                                                513217 non-null  object
 7   Incident ID                                           513217 non-null  object
 8   Incident Number                                       

In [50]:
newer_data['Incident Description'].unique()

array(['Malicious Mischief, Vandalism to Property', 'Arson',
       'Lost Property', 'Theft, From Locked Vehicle, >$950',
       'Suspicious Occurrence', 'Trespassing',
       'Vehicle, Recovered, Stolen outside SF', 'Mental Health Detention',
       'Burglary, Hot Prowl, Forcible Entry',
       'Theft, From Locked Vehicle, $200-$950',
       'Malicious Mischief, Vandalism to Vehicle', 'Battery, Sexual',
       'Firearm, Possession of Loaded', 'License Plate, Stolen',
       'Theft, From Unlocked Vehicle, >$950',
       'Weapon, Deadly, Imitation or Laser Scope, Exhibiting',
       'Vehicle, Stolen, Motorcycle', 'Investigative Detention',
       'Vehicle, Stolen, Auto', 'Vehicle, Recovered, Auto',
       'Found  Property', 'Theft, From Building, $200-$950',
       'Theft, Lost Property, Petty',
       'Firearm, Discharging Within City Limits',
       'Phone Calls, Harassing', 'Warrant Arrest, Local SF Warrant',
       'Burglary, Residence, Forcible Entry',
       'Burglary, Residence, 